### PR TITLE
Use reduced-size sample set for bootstrap estimation (delta, sobol)

### PR DESCRIPTION
**Pull request by simonmarwitz**
See Issue #570

### CODE REVIEW

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:18]: I propose to follow SciPy's naming here (that would be handy also if we do move towards using `scipy.stats.bootstrap`)

```suggestion
    n_resamples: int = None,

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:19]: I would call it `method` which could get `"all", "sobol", "delta"`. That would be more flexible for future improvements I think.

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:56]: Once we settle on the API, this needs a space in between.
```suggestion
    Nr : int

```

**Review by tupui — COMMENTED** [review]
Thank you for making a PR @simonmarwitz 🚀 

I left a few suggestions to get this started. We should also probably add some rudimentary tests. Since the changes are fairly "small", I am personally open to leave this for a follow up.

**Comment by simonmarwitz** [comment]
What do I have to do now? I suppose, it is up to me to implement your suggestions and somehow commit them here?!

**Comment by tupui** [comment]
So a typical workflow is that you would then add more commit to the same branch and push, which would update this PR.

This of course provided that @ConnectedSystems agrees with what I am proposing of course 😃 

For collaborative work, it's a bit difficult because I don't have write access to this repo  (yet? 😉). You could give explicit access to people to your fork otherwise. I am happy to guide/help if you need anything.

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [reply src/SALib/analyze/delta.py:18]: I think I need someone to explain to me conceptually why we need two separate arguments to do with resampling.

If it is indeed necessary to have both, `n_resamples` is too close to the current `num_resamples`, but I'd like to avoid capitalization in arguments where possible, so agree `Nr` has to be changed...

**Comment by ConnectedSystems** [comment]
Thanks @simonmarwitz , appreciate you contributing even while busy.

Please note my comments on your PR.

@tupui

> I don't have write access to this repo

Just sent you an invite!

File: src/SALib/analyze/delta.py
Comment (tupui) [reply src/SALib/analyze/delta.py:18]: Oh yes I got confused myself as well. `n_resamples` is not a good name here. Sorry for the confusion. I am not sure about naming here, maybe `n_subsamples`?

**Comment by tupui** [comment]
> @tupui
> 
> > I don't have write access to this repo
> 
> Just sent you an invite!

yay thanks Takuya! That's very exciting.

File: src/SALib/analyze/delta.py
Comment (simonmarwitz) [reply src/SALib/analyze/delta.py:18]: If you plan to use scipy.stats.bootstrap in the future, imho it would make sense to adopt the argument names `num_resamples` -> `n_resamples`

I would propose to name the size of the resample `Nr` -> `s_resamples`

File: src/SALib/analyze/delta.py
Comment (simonmarwitz) [reply src/SALib/analyze/delta.py:19]: I propose to completely remove sobol indices from this method and move it to a different file. It is simply not related to DMIM. Although a different implementation of sobol indices already exists in SALib.

File: src/SALib/analyze/delta.py
Comment (tupui) [reply src/SALib/analyze/delta.py:18]: We have a backward compatibility concern here, we sadly cannot rename `num_resamples -> n_resamples` without a deprecation cycle.

`s_resamples`, I am concerned this would be understood as `sobol_resamples`. Since we would reduce Y, why not just `y_resamples` (or `Y_resamples` but I am not a fan of the casing)?

**Comment by simonmarwitz** [comment]
I did some further research on the subject. Apparently here is a paper about it [https://arxiv.org/abs/1112.5016](url). They use resampling without replacement, though. I am not sure if that is easily to be implemented with randint indices. Also `len(Y)` would have to be sufficiently larger than `Nr` / `s_resamples` (what is "sufficiently larger" then?).

File: src/SALib/analyze/delta.py
Comment (simonmarwitz) [reply src/SALib/analyze/delta.py:18]: If you go for my proposal to remove sobol indices from this method, that would no longer be a problem.

Comment (tupui) [reply src/SALib/analyze/delta.py:18]: Please don't remove Sobol' indices from this method. We cannot do this like that for backward compatibility reasons. This is going to break people's code.

File: src/SALib/analyze/delta.py
Comment (tupui) [reply src/SALib/analyze/delta.py:19]: I replied above. But in the literature you do see expressions of Sobol' indices with moment independent. So this is just a different metric. I do agree that the name of the function is not general enough then to convey that. But we cannot change like this the parameters and features.

What we can do though is (in some follow up) to think about whether we would want or not to have a new API that would be more general. Doing so, we would be able to rework the API as we want and then ultimately deprecate this function.

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [reply src/SALib/analyze/delta.py:18]: I'm okay with `y_resamples`, seems like a good interim compromise. 

I'm also open to removing Sobol' (not in this PR, but in the future).

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:147]:
```suggestion
def bias_reduced_delta(Y, Ygrid, X, m, num_resamples, conf_level, y_resamples):

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:156]: I suppose we don't need to handle the None case as this function is more or less private.
```suggestion
    ind = np.random.randint(N, size= y_resamples)
    d_hat = calc_delta(Y[ind], Ygrid, X[ind], m)
    r = np.random.randint(N, size=(num_resamples, y_resamples))

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:185]:
```suggestion
def sobol_first_conf(Y, X, m, num_resamples, conf_level, y_resamples):

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:191]: Same here
```suggestion
    r = np.random.randint(N, size=(num_resamples, y_resamples))

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:58]:
```suggestion
    method : {"all", "delta", "sobol"}, optional

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:56]:
```suggestion
    y_resamples : int, optional

```

File: src/SALib/analyze/delta.py
Comment (tupui) [inline src/SALib/analyze/delta.py:102]:
```suggestion
            if method in ["all", "sobol"]:

```

**Review by tupui — CHANGES_REQUESTED** [review]
Thanks for the update! This looks in good shape besides a few tests that need to be updated. (I made suggestions to help 😃, you can go to the file tabs and group suggestions together.)

**Comment by ConnectedSystems** [comment]
I'll try to finalize comments/suggestions this week. Sorry for the delay in getting to this!

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [reply src/SALib/analyze/delta.py]: On second thought, I think best to leave the Sobol' estimation in (but optional).

I prefer `y_resamples`.

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [inline src/SALib/analyze/delta.py:58]: Could you make the spacing here consistent?

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [inline src/SALib/analyze/delta.py:82]: Remove extra space in error message ("total  number")

"less-equal" -> "less than or equal to"

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [inline src/SALib/analyze/delta.py:105]: Remove blank line here please

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [inline src/SALib/analyze/delta.py:151]: Please remove extra line here

File: src/SALib/analyze/delta.py
Comment (ConnectedSystems) [inline src/SALib/analyze/delta.py:189]: Same here, remove extra line

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Hi @simonmarwitz 

Thanks for these changes, I have a few more minor requests before I approve and merge.

Thanks again!

**Review by ConnectedSystems — APPROVED** [review]
Thanks @simonmarwitz !

**Comment by ConnectedSystems** [comment]
Hmm, some tests failing - I'll look into it soon.

**Comment by ConnectedSystems** [comment]
Apologies all, not sure what happened there as I definitely didn't mean to merge but the test fix is small.

**Pull request merged**
